cmd/dockerd/daemon.go

```diff
@@ -477,6 +477,16 @@ func (cli *DaemonCli) reloadConfig() {
 		}
 	}
 
+	// On Linux, we use sd_notify to indicate we're reloading config. We send
+	// this signal as early as possible to let systemd know we're reloading,
+	// but must signal "ready" after this completes (even on failure), which
+	// is done by the reload function defined above.
+	//
+	// We could make this more granular (i.e., don't notify if initial
+	// validation failed, and we aborted before actually reloading config.
+	go notifyReloading()
+	defer func() { go notifyReady() }()
```

Comment:
I made handling these (and the existing ones) asynchronously a choice for the caller, but perhaps we should consider indeed to not making these asynchronous; https://github.com/moby/moby/pull/47358#discussion_r1486179414

@neersighted @corhere WDYT?

Reply:
The intention seems to be avoiding blocking the caller with an "optional" task like this, but I wonder if this is too lightweight given there are now synchronization issues at play, and perhaps we should instead use a channel + background goroutine to coordinate these calls?

Reply:
Yeah, honestly not sure how heavy-weight these are. Perhaps it's fine to just remove the `go` here, and just call them synchronously 🤔

Reply:
I think it'd be fine to call them synchronously. [They open a UNIX datagram socket, write and close.](https://github.com/coreos/go-systemd/blob/d5623bf85e8e73ae6352f78ee6b55a287619dd4e/daemon/sdnotify.go#L48-L84) I highly doubt the calls would block for a significant length of time unless pid 1 has become unresponsive, in which case you'd have much bigger problems. It'd have to be lightweight and fast as I can't imagine many C programs would be going through the trouble of invoking `sd_notify()` in a background thread. (Note that the syslog API is very similar: apps log by sending unixgrams to the `/dev/log` socket.)